src/headerssync.h

```diff
@@ -0,0 +1,246 @@
+ * storing them in memory.
+ *
+ * To prevent an attacker from using (eg) the honest chain to convince us that
+ * they have a high-work chain, but then feeding us an alternate set of
+ * low-difficulty headers in the second phase, we store commitments to the
+ * chain we see in the first download phase that we check in the second phase,
+ * as follows:
+ *
+ * - In phase 1, store 1 bit (using a salted hash function) for every N headers
+ * that we see. With a reasonable choice of N, this uses relatively little
+ * memory even for a very long chain.
+ *
+ * - In phase 2 (redownload), keep a lookahead buffer of size H, and only
+ * accept a batch of N (N < H) headers to memory once we've verified that H/N =
+ * S commitments have all passed verification. With this parametrization, we
+ * can achieve a given security target (S) while choosing H and N to minimize
+ * memory usage in this scheme.
+ */
+
+class HeadersSyncState {
+public:
+    HeadersSyncState(NodeId id, const Consensus::Params& consensus_params);
+    ~HeadersSyncState() {}
+
+    enum class State {
+        /** INITIAL_DOWNLOAD means the peer has not yet demonstrated their
+         * chain has sufficient work */
+        INITIAL_DOWNLOAD,
+        /** REDOWNLOAD means the peer has given us a high-enough-work chain,
+         * and now we're redownloading the headers we saw before and trying to
+         * accept them */
+        REDOWNLOAD,
+        /** We're done syncing with this peer and can discard any remaining state */
+        FINAL
+    };
+
+    /** Return the current state of our download */
+    State GetState() const { return m_download_state; }
+
+    /** Start headers sync (via this download-twice mechanism)
+     * chain_start: best known fork point that the peer's headers branch from
+     * initial_headers: first batch of headers to process
+     * minimum_required_work: amount of chain work required to accept the chain
+     */
+    std::optional<CBlockLocator> StartInitialDownload(const CBlockIndex* chain_start, const
+            std::vector<CBlockHeader>& initial_headers, const arith_uint256&
+            minimum_required_work, CBlockLocator&& chain_start_locator);
+
+    /** Process a batch of headers, once a sync via this mechanism has started
+     *
+     * headers: headers that were received over the network for processing
+     * full_headers_message: true if the message was at max capacity,
+     *                       indicating more headers may be available
+     * headers_to_process: will be filled in with any headers that the caller
+     *                     can process and validate now (because these returned
+     *                     headers are on a chain with sufficient work)
+     * processing_success: set to false if an error is detected and the sync is
+     *                     aborted; true otherwise.
+     */
+    std::optional<CBlockLocator> ProcessNextHeaders(const std::vector<CBlockHeader>& headers,
```

Comment:
In commit "Utilize anti-DoS headers download strategy"

This function really returns 3 things: an optional block locator, a vector of headers to process, and a bool processing_success. Having those spread over return values and mutable arguments is a bit ugly.

How about returning a typedef'd `std::optional<std::pair<std::optional<CBlockLocator>, std::vector<CBlockHeader>>>`, or making a simple custom struct to return the results in?

Reply:
This is way better, thanks.